### ISSUE TITLE
[Source Request]: Heidelberg with new web app

### ISSUE THREAD

**Issue by DerDreschner** [mampfes/hacs_waste_collection_schedule#3248]
### Municipality / Region

Heidelberg, Germany

### Collection Calendar Webpage

https://abfallkalender.heidelberg.de/

### Example Address

Albert-Fritz-Straße Nr. 1-59, 2-66, 69124 Heidelberg

### Collection Data Format

Something different (add to additional information)

### Additional Information

Heidelberg stopped using gipsprojekt this week. Instead, they have a new web app under [abfallkalender.heidelberg.de](https://github.com/mampfes/hacs_waste_collection_schedule/issues/abfallkalender.heidelberg.de). The app generates an ICS file by itself. The data comes from these two web services:

https://garbage.datenplattform.heidelberg.de/streetnames
[https://garbage.datenplattform.heidelberg.de/collections?street={streetname}](https://garbage.datenplattform.heidelberg.de/collections?street=%7Bstreetname%7D)

The data itself isn't very helpful. It has to be processed to know the exact waste collection dates.

Property "calendar_week_{waste_type}":
"A" means collection on all weeks
"G" means all uneven weeks
"U" all even weeks
...on the day specified in "day_of_week_{waste_type}" (german day name in short!)
-> except when the day would fall on a holiday ("holidays" property), then the information about when the collection is happening is inside the "exceptions" property

Christmas tree collection is specified in the seperate "christmas" property.

**Comment by jandechent** [comment]
Yeah, Heidelberg broke. There is a ics download button - but this calls some lengthy js code to provide the ics file. The question is now: Can we make hacs_waste_collection_schedule call the js function to get the file, or do we have to replicate what the JS code does to get to the ics?

**Comment by jandechent** [comment]
Aside the streetname, the user would have to choose how they get their waste collected. This would result in the following options:

 - Rest: Wöchentlich/Bedarf oder 14-tägiglich
 - Bio: Wöchentlich oder keine Abholung
 - Papier: Wöchentlich oder 14-tägiglich
 - Gelb: Wöchentlich oder 14-tägiglich

With that, creating a calendar should be very possible. The exceptions cover all days, even those, where in that street no collection takes place.  

Does someone have a example source, that I could try to adapt to make this work?

**Comment by DerDreschner** [comment]
@jandechent : I am working on an implementation right now.

Honestly, selecting the collection schedule makes only sense for residual/general waste (Restmüll) in our case.

- Residual waste: If collected bi-weekly, it's important to know if your house number is even or not. In that case, they alternate every week between even and uneven house numbers (street sides). It's the only case where the house number is important.
- Garden waste: Isn't relevant in our case, as the script shouldn't limit the returned waste types (is collected weekly or should be hidden by the user in HA later).
- Paper waste: Is only collected on a biweekly basis by the city. Exceptions are only possible for commercial customers in selected streets ([Papierbündelsammlung](https://www.heidelberg.de/HD/Rathaus/Papierentsorgung+fuer+Gewerbe.html)). Therefore, I would completely ignore that case.
- Recyclable waste (Gelbe Tonne): Is only collected on a biweekly basis by the city. I couldn't find any exceptions to that schedule. No idea why it's possible to select a weekly schedule here? I will ask the city by mail later.

Let me know if you have any notes/thoughts.

**Comment by jandechent** [comment]
That sounds great and correct. Tell me if you need help and lets hope Home Assistant repairs the calendar integration fast.

**Comment by DerDreschner** [comment]
@jandechent : Great. I've done with my implementation and opened PR #3477 . As I'm usually working with C# and PHP, I may have used Python a bit wrong. A code review or any other feedback is therefore very much appreciated.

**Comment by tkaefer** [comment]
Works like charm for me - thank you very much.
